### PR TITLE
[Feature] Web Flow Tests [CAS-66]

### PR DESCRIPTION
**Pull request by cslzchen**
### Purpose

Add tests for web login flow action `remoteAuthenticate` and class `OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction`.

- [x] Institution Login with SAML by Shibboleth.
- [x] Institution Login with CAS by pac4j.
- [x] ORCid Login with OAuth by pac4j.
- [x] OSF Login with `username` and `verification_key`

### Tickets

https://openscience.atlassian.net/browse/CAS-66

### CODE REVIEW

File: cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java [hunk 1]
```diff
@@ -106,7 +106,7 @@
  * @author Longze Chen
  * @since 4.1.5
  */
-public final class OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction
+public class OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java:109]: Must not be `final` to be extended by mock classes.

File: cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java [hunk 2]
```diff
@@ -384,7 +388,7 @@ protected OpenScienceFrameworkCredential constructCredential(final RequestContex
      * @return the username from the idp and setup on the OSF
      * @throws AccountException a account exception
      */
-    private PrincipalAuthenticationResult notifyRemotePrincipalAuthenticated(
+     protected PrincipalAuthenticationResult notifyRemotePrincipalAuthenticated(
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java:391]: Must not be `private` for its method to be overridden by sub-classes.

File: cas-server-support-osf/src/test/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsActionTests.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+    private static final String PAC4J_DELEGATION_PROFILE_ID = "MockCasProfile#0001-1234-5678";
+
+    @Test
+    public void verifyInstitutionSamlShibbolethFlow() throws Exception {
+        assertTrue(Boolean.TRUE);
+    }
+
+    @Test
+    public void verifyInstitutionCasPac4jFlow() throws Exception {
```
Comment (cslzchen) [inline cas-server-support-osf/src/test/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsActionTests.java]: For institution login by CAS pac4j, verify the final `credential` and the return event.
TODO: add check that `TGT` is destroyed.

File: cas-server-support-osf/src/test/java/io/cos/cas/mock/MockOsfRemoteAuthenticateAction.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+import javax.security.auth.login.AccountException;
+
+/**
+ * This class mocks the {@link OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction} class.
+ *
+ * @author Longze Chen
+ * @since  4.1.5
+ */
+public class MockOsfRemoteAuthenticateAction
```
Comment (cslzchen) [inline cas-server-support-osf/src/test/java/io/cos/cas/mock/MockOsfRemoteAuthenticateAction.java:16]: Need to mock and extend the class `OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction` to override the method `notifyRemotePrincipalAuthenticated` that talks to OSF API.

**Review by cslzchen — COMMENTED** [review]
WIP but ready for advice @icereval.

**Review by icereval — APPROVED** [review]

**Comment by cslzchen** [comment]
All 4 tests are done. Ready for review/merge when tests passed.

**Review by icereval — APPROVED** [review]
Thanks!

**Pull request merged**